src/devices/machine/i8291a.cpp

```diff
@@ -0,0 +1,1082 @@
+#define LOG_CMD_MASK	(LOG_INT_MASK << 1)
+#define LOG_CMD(...)	LOGMASKED(LOG_CMD_MASK, __VA_ARGS__)
+#undef VERBOSE
+#define VERBOSE (LOG_GENERAL)//|LOG_REG_MASK|LOG_CMD_MASK|LOG_STATE_MASK|LOG_STATE_MASK)
+
+DEFINE_DEVICE_TYPE(I8291A, i8291a_device, "i8291a", "I8291A")
+
+i8291a_device::i8291a_device(const machine_config &mconfig, const char *tag,
+		device_t *owner, uint32_t clock) :
+		device_t(mconfig, I8291A, tag, owner, clock),
+		m_int_write_func(*this),
+		m_dreq_write_func(*this),
+		m_trig_write_func(*this),
+		m_eoi_write_func{*this},
+		m_dav_write_func{*this},
+		m_nrfd_write_func{*this},
+		m_ndac_write_func{*this},
+		m_srq_write_func{*this},
+		m_dio_write_func{*this},
+		m_dio_read_func{*this}
+				{
+}
+
+void i8291a_device::device_reset()
+{
+	m_pon = true;
+	m_ints1 = 0;
+	m_ints2 = 0;
+	m_auxa = 0;
+	m_auxb = 0;
+	m_spoll_mode = 0;
+	// TODO: clear parallel poll
+	m_address_status &= ~REG_ADDRESS_STATUS_EOI;
+	m_rdy = true;
+	m_nba = false;
+	m_send_eoi = false;
+	m_ist = false;
+	m_eoi = false;
+	m_ifc = false;
+	m_srq = false;
+	m_dav = false;
+	m_ndac = false;
+	m_nrfd = false;
+	m_atn = false;
+	m_ren = false;
+	update_state(m_t_state, TalkerState::TIDS);
+	update_state(m_tp_state, TalkerPrimaryState::TPIS);
+	update_state(m_tsp_state, TalkerSerialPollState::SPIS);
+	update_state(m_l_state, ListenerState::LIDS);
+	update_state(m_rl_state, RemoteLocalState::LOCS);
+	update_state(m_pp_state, ParallelPollState::PPIS);
+	update_state(m_dc_state, DeviceClearState::DCIS);
+	update_state(m_dt_state, DeviceTriggerState::DTIS);
+	update_state(m_sh_state, SourceHandshake::SIDS);
+	update_state(m_ah_state, AcceptorHandshake::AIDS);
+}
+
+void i8291a_device::device_start()
+{
+	m_int_write_func.resolve_safe();
+	m_dreq_write_func.resolve_safe();
+	m_dav_write_func.resolve_safe();
+	m_nrfd_write_func.resolve_safe();
+	m_ndac_write_func.resolve_safe();
+	m_eoi_write_func.resolve_safe();
+	m_srq_write_func.resolve_safe();
+	m_dio_read_func.resolve_safe(0xff);
+	m_dio_write_func.resolve_safe();
+
+}
+
+READ8_MEMBER(i8291a_device::din_r)
+{
+	LOG_REG("%s: %02X\n", __FUNCTION__, m_din);
+	m_din_flag = false;
+	m_ints1 &= ~REG_INTS1_BI;
+	update_int();
+	run_fsm();
+	return m_din;
+}
+
+void i8291a_device::update_int()
+{
+	bool dreq = ((m_ie2 & REG_IE2_DMAO) && (m_ints1 & REG_INTS1_BO)) |
+			((m_ie2 & REG_IE2_DMAI)  && (m_ints1 & REG_INTS1_BI));
+
+	if (m_dreq_out != dreq) {
+		LOG_INT("%s: dreq %s\n", __FUNCTION__, dreq ? "true" : "false");
+		m_dreq_write_func(dreq);
+		m_dreq_out = dreq;
+	}
+
+	bool intr = false;
+
+	if (m_ints1 & m_ie1)
+		intr = true;
+
+	if ((m_ints2 & m_ie2) & 0x0f)
+		intr = true;
+
+	if (intr) {
+		m_ints2 |= REG_INTS2_INT;
+		m_address0 |= REG_ADDRESS0_INT;
+	} else {
+		m_ints2 &= ~REG_INTS2_INT;
+		m_address0 &= ~REG_ADDRESS0_INT;
+	}
+
+	if (m_intr_out != intr) {
+		LOG_INT("%s: intr %s\n", __FUNCTION__, intr ? "true" : "false");
+		m_int_write_func(intr);
+		m_intr_out = intr;
+	}
+
+
+
+}
```

Comment:
What's the big vertical whitespace inside the function supposed to be in aid of?